Stop autoforward on BACKWARD key-press (if autoforward is active)

tukkek opened the pull request:
Hi! I love the autoforward feature so I thought I would enhance it with this behavior. It comes naturally for me from playing other games with autorun in them. Also, since you cannot possibly walk forward and backward at the same time, it's only natural that pressing backwards would stop you moving forward, I believe.

I was not sure if I should create a new branch for this or not but since this is such a small commit, I thought I'd just send it to `master` anyway. Let me know if I should create a new request and thanks for your consideration!

tukkek commented on src/game.cpp:
```diff
@@ -1861,6 +1861,10 @@ void Game::processKeyInput()
+	} else if (wasKeyDown(KeyType::BACKWARD)) {
+                if(g_settings->getBool("continuous_forward")){
+                    toggleAutoforward();
+                }
```
I used the `if` statement just below as a reference in deciding whether to keep braces or not https://github.com/tukkek/minetest/blob/c2ae73f0ba2cc92adf74100e26892afbab9e4a23/src/game.cpp#L1871

sfan5 replied:
http://dev.minetest.net/Code_style_guidelines

tukkek commented:
@sfan5 I have made the style changes you asked. Thanks for the feedback.

(by the way the link you sent me says only same-line `if` statements are disallowed, it says nothing in particular about braces - you may want to update it if that's a real project guideline)

SmallJoker commented on src/game.cpp:
```diff
@@ -1861,6 +1861,9 @@ void Game::processKeyInput()
 	} else if (wasKeyDown(KeyType::AUTOFORWARD)) {
 		toggleAutoforward();
+	} else if (wasKeyDown(KeyType::BACKWARD)) {
+		if(g_settings->getBool("continuous_forward"))
```
Missing space after `if`

SmallJoker commented:
@tukkek Our C(++) code style guidelines are based on the Linux kernel style, which clearly states that brackets for a single line after a condition should be removed.

> Do not unnecessarily use braces where a single statement will do.
https://www.kernel.org/doc/html/latest/process/coding-style.html#placing-braces-and-spaces

tukkek commented:
I have fixed the missing whitespace as per @SmallJoker's comment and I have squashed the commits to prevent log spam as well. Thanks for the feedback!

nerzhul commented:
strange, if i press backward and continuous is enabled, and i'm not doing it, it will enable it no ?

i think backward should cancel it ONLY if it's currently enabled (not the setting but the feature)

rubenwardy commented:
is `wasKeyDown` the one with side effects?

nerzhul commented:
note: i just expose a theory based on the current PR algorithm, i didn't tested it, but logically it should enable automove when pressing backward and not moving

tukkek commented:
> Don't merge yet, back key already stops autoforward.

Is this recently added? Because on my end it doesn't (otherwise I wouldn't have coded it myself).

As for toggling behavior: read title. As for side-effects: I'm using the same function as all other input-handling events.

paramat commented:
No not recent at all, maybe 2 years ago. I have read the title and this happens already. What key do you assign for autoforward?

paramat commented:
Let's see how this behaves for others. I do support the feature beause it happens for me and it's useful.
Maybe it's a keyboard hardware issue.

paramat commented:
Ah, do you mean stop motion on back key press instead of back key hold? Looks like it from the code.
For me holding back key stops autoforward, which seems good. I'm not sure about a back key press-and-release toggling autoforward because you use the autoforward key for that, but maybe.

paramat commented:
But then, will this PR make back key toggle autoforward? Looks like it.

tukkek commented:
I'm sorry my friend but you are wrong. I have just now checked all occurrences of when the game sets the `continuous_forward` setting and there's simply no other situation where autorun is disabled except when the autoforward key is pressed. In other words: **pressing back does not disable autorun** in the current game.

I also don't think autorun was available in the game as of 2 years ago, as you claim.

If you have proof of what you are saying (and by that I mean a file and code line where this is achieved), feel free to show it to us.

Also: **back key does not toggle autoforward**, read the title.

nerzhul commented:
@tukkek we read the title, but i read the code, and reading our code it's unclear, i see we read a setting and toggle the flag :)

tukkek commented:
Is it possible you are using a auto-forward mod @paramat? This change is about the base-game feature. This seems to be the only possible explanation regarding what you are saying.

Again: feel free to show us in the game's code where you claim that backwards will disable auto-forward.

tukkek commented:
> we read a setting and toggle the flag :)

Exactly: the toggle is only done if autorun is enabled. As stated very clearly in the title and as I have explained to everyone here three times already.

paramat commented:
```
// Autoforward by toggling continuous forward.
void Game::toggleAutoforward()
{
	bool autorun_enabled = !g_settings->getBool("continuous_forward");
	g_settings->set("continuous_forward", bool_to_cstr(autorun_enabled));
```
The setting does not enable autoforward, it actually controls it, the setting flips each time you press the autoforward key.
I personally worked on the autoforward PRs and have used the feature for a while, 2 years may be wrong, check commits, maybe 1 year.
> I have just now checked all occurrences of when the game sets the continuous_forward setting and there's simply no other situation where autorun is disabled except when the autoforward key is pressed.

Correct, as i wrote, maybe it's a hardware issue and maybe it doesn't happen to everyone.

paramat commented:
Oops i'm sorry, misread code, i'm tired :)
Indeed it will not toggle autoforward on.

tukkek commented:
I'm still waiting for you to show us where pressing back disables autorun. It's not a hardware issue, it's a simple matter that there isn't any code in the project that does that. You can see if very simply by doing a code search for usage of `continuous_forward` or `toggleAutoforward()`.

Please stop spamming here until you can actually prove what you're saying. Repeating your claim and posting random snippets of code that doesn't answer the quesiton isn't going to help anyone.

paramat commented:
Anyway, does holding back key stop motion for you? What about others?
But still, i support the feature, useful.

tukkek commented:
Yes, it stops the motion but it doesn't stop autoforward.

paramat commented:
> I'm still waiting for you to show us where pressing back disables autorun.

As i wrote above, there is no code that does that, you are correct. I'm just curioous why this happens for me and if it happens for others, because it is closely related.
No need to be rude.
I will very probably +1 this PR.

tukkek commented:
I'm not being rude, I'm just making abundantly clear for the fourth time that:

* **This patch only disables auto-forward if auto-forward is currently active.**
* **Autoforward is currently being disabled only if you press the autoforward key again.**

Sorry if I sound rude - I don't mean to be. I just wasn't expecting a long-winded discussion over literally 3 lines of code, but I do apologize if I'm deliberately being very direct.

Have you checked whether you have an autorun mod installed instead? That might be it - or maybe some other mod that messes with your settings when your press a key?

paramat commented:
> Stop autoforward on BACKWARD key-press (if autoforward is active)

That can easily be interpreted as meaning autoforward is temporarily stopped on back key hold, which happens already for you and me. This caused the confusion. Maybe other devs were not aware that holding back key temporarily stopped movement, if they were aware they would quite possibly write what i did.
I removed the quoted code above because i was wrong, and am sorry for that.
No i'm not using an autoforward mod :)

Anyway once i test i will probably +1 this.

paramat commented:
Yes it's clear now don't worry.

tukkek commented:
> This caused the confusion

Got it! Glad this was cleared up ^^ thank you for your support.

paramat commented:
:+1: Tested.

tukkek commented:
@paramat I don't think your review came through, GitHub is only showing 1 approval for me.

Since you're all project members, can you inform me whether is there anything else I should do before this is merged? Thanks!

paramat commented:
That's probably because i don't use the Github review feature, i review as we did before, where a :+1: in a comment means an official approval and the official number of approvals are shown in the thread labels.

paramat commented:
Hmm maybe document it somewhere? Looking for where ..

tukkek commented:
I have created the documentation update on a different pull request (#7417) since this one is already resolved with two approved reviews. What's the next step towards getting this specific pull request merged?

paramat commented:
The docs can be added here, the approvals don't prevent that, we'll just check the docs are ok.

tukkek commented:
Please do that on the other pull request then since it contains both docs and code patch (full package), thanks! https://github.com/minetest/minetest/pull/7417

tukkek commented:
README.md makes no mention of autoforward so I don't think it needs to be updated.

paramat commented:
Replaced by #7417
